[BugFix] Use async Mistral Tokenizer in Chat Completions

bbrowning opened the pull request:
## Purpose

When using `--tokenizer-mode mistral` with Chat Completions, large inputs could block the event loop, causing delays in processing of other requests. This adjusts the usage of that tokenizer during chat request preprocessing to use pre-existing but unused `_tokenizer_executor` in OpenAIServing to run these blocking operations in a background thread as opposed to running them directly in the critical path of the server event loop.

Fixes #24910 

## Test Plan

I added a new `tests/entrypoints/openai/test_serving_engine.py` with a single test that ensures the newly extracted `_apply_chat_template` method properly offloads the long-running tokenization operation to a background thread and does not block the event loop. For testing simplicity, this mocks the actual tokenization to just sleep to get a consistent simulation of a long tokenization.

To run:

```
 pytest -sv tests/entrypoints/openai/test_serving_engine.py
```

## Test Result

I ensured the test failed before I moved the execution into the threadpool executor. And, after moving it into the executor, the test passes. The test doesn't count the event loop as blocked until it's been blocked for at least 0.5 seconds, which is far higher than required in most scenarios but a value that large was chosen to ensure this doesn't flake on a heavily loaded or slow CI machine.

bbrowning committed:
[BugFix] Use async Mistral Tokenizer in Chat Completions

When using `--tokenizer-mode mistral` with Chat Completions, large
inputs could block the event loop, causing delays in processing of other
requests. This adjusts the usage of that tokenizer during chat request
preprocessing to use pre-existing but unused `_tokenizer_executor` in
OpenAIServing to run these blocking operations in a background thread as
opposed to running them directly in the critical path of the server
event loop.

Signed-off-by: Ben Browning <bbrownin@redhat.com>

bbrowning commented:
For a visual overview of the changes here, below is a before and after sequence diagram. This does not change the time a single large request takes when using `--tokenizer mistral`. If that took 50 seconds before, it will still take 50 seconds. All this does is free up the web server's event loop for processing other requests during that 50 seconds, which allows things like requests to `/health` or other APIs to respond quickly. If two large input requests come in that each take 50 seconds, those would take 100 seconds to complete before and will still take 100 seconds to complete now.

In other words, this isn't adding any parallelization of the Mistral tokenizer usage. That could be something considered as a separate PR depending on community and user needs for concurrent request performance with large inference inputs with that tokenizer.

## Before

```mermaid
  sequenceDiagram
      participant U1 as User 1
      participant U2 as User 2 (Health Check)
      participant EL as Event Loop
      participant MT as Mistral Tokenizer

      Note over U1,MT: BEFORE: Blocking Tokenization
      U1->>+EL: POST /v1/chat/completions (large input)
      EL->>MT: apply_mistral_chat_template
      activate MT
      Note over MT: 50 seconds processing...

      U2->>EL: GET /health
      Note over EL,U2: Request queued, waiting for event loop

      MT-->>EL: Tokenization complete
      deactivate MT
      EL-->>-U1: Chat Response (after 50s)

      EL-->>U2: Health OK (delayed ~50s)
```

## After

```mermaid
  sequenceDiagram
      participant U1 as User 1
      participant U2 as User 2 (Health Check)
      participant EL as Event Loop
      participant BG as Background Thread
      participant MT as Mistral Tokenizer
      
      Note over U1,BG: AFTER: Async Tokenization
      U1->>+EL: POST /v1/chat/completions (large input)
      EL->>BG: async_apply_mistral_chat_template
      activate BG
      deactivate EL

      BG->>+MT: apply_mistral_chat_template
      Note over MT: 50 seconds processing...

      U2->>+EL: GET /health
      Note over EL: Event loop available!
      EL-->>-U2: Health OK (immediate response)

      MT-->-BG: Tokenization complete
      BG-->>U1: Chat Response (after 50s)
      deactivate BG
```

bbrowning commented:
I'm going to mark this as draft, as I see some in-progress work at #22880 that's starting to merge and will impact whether this is still an issue and if so how we fix it. Specifically, the issue described at #23873 as part of that overall work will change this. So, will revisit once some of that lands unless there's a strong desire to preempt that work.

I have also since discovered the existing `make_async` method, so even if we did want to land just this fix I can simplify the implementation.

bbrowning committed:
Refactor to use make_async utils method

After digging a bit more, I discovered we have an existing `make_async`
method that encapsulates turning a block call into an async one. So, I'm
simplifying this change to take advantage of that new method. In doing
so, I needed to pull the if statement that determines which chat
template to apply out of `_preprocess_chat` and into its own
`_apply_chat_template` to make testing of this newly simplified path
easier without having to do a lot of mocking for all the other concerns
covered in the larger `_preprocess_chat` method.

Signed-off-by: Ben Browning <bbrownin@redhat.com>

bbrowning commented:
Updated to use `make_async`, pulling out of draft so that others can decide if this should merge sooner or wait on the larger refactoring of chat preprocessing.

DarkLight1337 reviewed (approved):
The Renderer refactor is going to take a while, let's have this bandaid fix for now.

DarkLight1337 committed:
Merge branch 'main' into 20251002-mistral-tokenizer-async

njhill commented on vllm/entrypoints/openai/serving_engine.py:
```diff
@@ -793,23 +819,12 @@ async def _preprocess_chat(
 
-        request_prompt: Union[str, list[int]]
-
-        if tokenizer is None:
-            request_prompt = "placeholder"
```
we could still do the non-tokenizer case inline to avoid the await barrier in this case...

njhill commented on vllm/entrypoints/openai/serving_engine.py:
```diff
@@ -749,6 +749,32 @@ async def _tokenize_prompt_inputs_async(
                     tokenizer=tokenizer,
                 )
 
+    async def _apply_chat_template(
+            self, tokenizer: AnyTokenizer,
+            messages: list[ChatCompletionMessageParam],
+            conversation: list[ConversationMessage],
+            chat_template_kwargs: dict[str, Any]) -> Union[str, list[int]]:
+        request_prompt: Union[str, list[int]]
+
+        if tokenizer is None:
+            request_prompt = "placeholder"
+        elif isinstance(tokenizer, MistralTokenizer):
+            apply_mistral_async = make_async(apply_mistral_chat_template,
+                                             executor=self._tokenizer_executor)
+            request_prompt = await apply_mistral_async(
+                tokenizer,
+                messages=messages,
+                **chat_template_kwargs,
+            )
+        else:
+            request_prompt = apply_hf_chat_template(
+                tokenizer=tokenizer,
+                conversation=conversation,
+                model_config=self.model_config,
+                **chat_template_kwargs,
+            )
+        return request_prompt
```
How about keep this as a non-async method and in the constructor have:
```python
self._apply_chat_template_async = make_async(self._apply_chat_template,
                                             executor=self._tokenizer_executor)
```
so that
- we do same thing for non-mistral case (which in theory could similarly block the event loop?)
- we don't create the async function every time

bbrowning replied:
My understanding of the code is we don't want to do this for the non-mistral case, as in `apply_hf_chat_template` looks to be non-blocking and doesn't actually do any tokenization as part of applying its chat template. For that code path, the blocking tokenization actually gets handled further down in `_preprocess_chat ` in code like:

```python
        elif isinstance(request_prompt, str):
            prompt_inputs = await self._tokenize_prompt_input_async(
                request,
                tokenizer,
                request_prompt,
                add_special_tokens=add_special_tokens,
            )
```

There, the blocking work for the HF tokenizer path already happens in the specialized `AsyncMicrobatchTokenizer`, although it takes a few levels of indirection to get there.

However, `apply_mistral_chat_template` does actually do tokenization, returning a list of token ids. So, that needs to happen async but cannot use the existing `AsyncMicrobatchTokenizer` because it relies on code inside the mistral library to do this tokenization.

Your point about not recreating the async function every time is valid, and I can hoist that up to an instance variable in the class to hold this async version of the method.

As far as avoiding the await barrier, I refactored this chat template application logic into its own method just to make testing simpler. I can unravel a bit of that, so that we only cross the await barrier in the case of the mistral tokenizer here if that's preferred.

njhill reviewed (commented):
Thanks @bbrowning! I have a couple of minor suggestions, also fine to merge this and handle in a follow-on PR...

bbrowning committed:
Further simplify async mistral tokenizer fix

This reduces the scope of the changes to just the bare minimum required
to get the `apply_mistral_chat_template` running in a background thread
without any other changes, to avoid introducing any additional `await`
barriers into hot code paths for requests not using mistral tokenizers.

Signed-off-by: Ben Browning <bbrownin@redhat.com>
